src/core/Akka.Streams/Implementation/Fusing/ActorGraphInterpreter.cs

```diff
@@ -872,26 +908,41 @@ public static Props Props(GraphInterpreterShell shell)
             return Actor.Props.Create(() => new ActorGraphInterpreter(shell)).WithDeploy(Deploy.Local);
         }
 
-        private readonly ISet<GraphInterpreterShell> _activeInterpreters = new HashSet<GraphInterpreterShell>();
+        private ISet<GraphInterpreterShell> _activeInterpreters = new HashSet<GraphInterpreterShell>();
         private readonly Queue<GraphInterpreterShell> _newShells = new Queue<GraphInterpreterShell>();
         private readonly SubFusingActorMaterializerImpl _subFusingMaterializerImpl;
         private readonly GraphInterpreterShell _initial;
         private ILoggingAdapter _log;
+        //this limits number of messages that can be processed synchronously during one actor receive.
+        private readonly int _eventLimit;
+        private int _currentLimit;
+        //this is a var in order to save the allocation when no short-circuiting actually happens
+        private Queue<object> _shortCircuitBuffer;
 
         public ActorGraphInterpreter(GraphInterpreterShell shell)
         {
             _initial = shell;
 
             _subFusingMaterializerImpl = new SubFusingActorMaterializerImpl(shell.Materializer, RegisterShell);
+            _eventLimit = _initial.Materializer.Settings.SyncProcessingLimit;
+            _currentLimit = _eventLimit;
         }
 
         public ILoggingAdapter Log => _log ?? (_log = Context.GetLogger());
 
+        private void EnqueueToShortCircuit(object input)
+        {
+            if(_shortCircuitBuffer == null)
+                _shortCircuitBuffer = new Queue<object>();
```

Comment:
I see no reason for a LinkedList and Queue has the simpler API for that use case. 

```var element = queue.Dequeue()``` 

vs
```
var element = list.First;
list.RemoveFirst();
```